src/kg_gen/utils/llm_deduplicate.py

```diff
@@ -0,0 +1,305 @@
+from scipy.spatial.distance import cdist
+import faiss
+from concurrent.futures import ThreadPoolExecutor
+import dspy
+from kg_gen.models import Graph
+import logging
+from sklearn.metrics.pairwise import cosine_similarity
+from rank_bm25 import BM25Okapi
+from sentence_transformers import SentenceTransformer
+import numpy as np
+import os
+
+
+class LLMDeduplicate:
+    graph: Graph
+    nodes: list[str]
+    edges: list[str]
+    node_clusters: list[list[str]]
+    edge_clusters: list[list[str]]
+
+    logger: logging.Logger = logging.getLogger(__name__)
+
+    def __init__(self, graph: Graph):
+        """
+        Initialize KG-assisted RAG with cached embeddings, BM25 tokens, and text chunk store.
+        """
+        self.graph = graph
+        self.nodes = list(graph.entities)
+        self.edges = list(graph.edges)
+        self.node_clusters = graph.entity_clusters or []
+        self.edge_clusters = graph.edge_clusters or []
+
+        # Embeddings and BM25 tokens for nodes
+        self.node_encoder = SentenceTransformer(
+            'sentence-transformers/all-mpnet-base-v2')
```

Comment:
ideally the models are configurable. not sure their default can be the same as the retriever model?

what properties of this model should we be looking for? i have yet to start playing with this code, but the nodes and edges below are "few words, not even sentences" strings, right?

i have multilingual requirements, so i need to investigate what is best suitable. (the inflect usage is another issue).

in any case @reinier-millo thanks a lot for providing this; it's defintely a reference implementation we can build on!

Reply:
> ideally the models are configurable. not sure their default can be the same as the retriever model?
> 

Yes, for sure. Just used that model that was the one used by the dedup algorithm. Added a dropdown setting for selecting the retrieval model and allow to use different models. Just added few of them in the list.


> what properties of this model should we be looking for? i have yet to start playing with this code, but the nodes and edges below are "few words, not even sentences" strings, right?
> 

Yes, nodes and edges are basically few words.

> i have multilingual requirements, so i need to investigate what is best suitable. (the inflect usage is another issue).
> 

Yes, for sure we will need to include multilingual models.